proposals/data-classes.md

```diff
@@ -0,0 +1,455 @@
+*nominal records*, in contrast to the existing [records proposal](https://github.com/dotnet/csharplang/blob/master/proposals/records.md), which I'll
+refer to as *positional records*. Nominal records are meant to satisfy the
+goals listed above by doing the following things:
+
+1. Automatically generating `Equals`, `GetHashCode`, `ToString`, `==`, `!=`, and `IEquatable<T>`
+   based on the member data of the type.
+1. Allow object initializers to also initialize readonly members.
+
+The LoginResource class now could be defined as
+
+```C#
+public data class LoginResource
+{
+    public string Username { get; }
+    public string Password { get; }
+    public bool RememberMe { get; } = false;
+}
+```
+
+and the use would be identical:
+
+```C#
+var x = new LoginResource {
+    Username = "andy",
+    Password = password
+};
+```
+
+Note that `RememberMe` must have an initializer to avoid a warning in the
+object initializer about an unset read-only property.
+
+**Why have nominal records?**
+
+Nominal records provide two major use cases that are not addressed by positional
+records. The first is adoption convenience. The most common use of a
+record-like type in C# today is a simple class with an object initializer.
+Transforming this into a positional record would require replacing all fields
+with parameters on the primary constructor, replacing all uses of object
+initializers with constructor calls, and explicitly providing default values
+for all parameters which are optional (where this may have been implicitly
+supplied before by simply not putting an initializer on a field or property).
+
+The second, more important use case is API evolution. Positional records make
+it difficult to add or re-order members without producing a breaking change.
+Reordering is essentially impossible for positional records without
+introducing a binary breaking change. Adding new members is difficult and
+requires manual intervention to create constructor overloads with the old
+members which must be carefully maintained. For many uses of records, like
+"options containers" (see
+[Roslyn](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/CompilationOptions.cs,29a3818343614151),
+[Benchmark.NET](https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/ConsoleArguments/CommandLineOptions.cs)
+for examples), this is a headache. Often data should not have positional
+dependence, and retrofitting independence later is quite difficult.
+
+### Code generation
+
+Aside from places where semantics are underspecified, this section is mostly
+for edification. The implementation of the feature is irrelevant to the
+language design itself. The implementation may change significantly as it is
+actually developed.
+
+That being said, the current proposal is that the generated class code would
+look like:
+
+```C#
+public class LoginResource : IEquatable<LoginResource>
```

Comment:
Can we make IEquatable impl opt-in? e.g. only if spelled out in the source: `data class C : IEquatable<C>`

Reply:
to reduce the amount of possibly unused code generated. the opt-in solution would be synonymous to `derive` or `deriving` in rust and haskell.